pennylane/io/qualtran_io.py

```diff
@@ -19,22 +19,187 @@
 # pylint: disable=unused-argument
 
 from collections import defaultdict
-from functools import lru_cache, singledispatch
+from functools import cached_property, singledispatch
+from typing import Dict, List, Tuple
 
 import numpy as np
 
+import pennylane.measurements as qmeas
 import pennylane.ops as qops
-from pennylane.operation import DecompositionUndefinedError, MatrixUndefinedError, Operation
+import pennylane.templates as qtemps
+from pennylane.operation import (
+    DecompositionUndefinedError,
+    MatrixUndefinedError,
+    Operation,
+    Operator,
+)
 from pennylane.registers import registers
+from pennylane.tape import make_qscript
 from pennylane.wires import WiresLike
+from pennylane.workflow import construct_tape
+from pennylane.workflow.qnode import QNode
 
 try:
+    import cirq
     import qualtran as qt
+    from qualtran import Bloq
+    from qualtran._infra.gate_with_registers import split_qubits
+    from qualtran.bloqs import basic_gates as qt_gates
+
+    qualtran = True
 except (ModuleNotFoundError, ImportError) as import_error:
-    pass
+    qualtran = False
+
+    Bloq = object
+
+
+@singledispatch
+def _get_op_call_graph(op):
+    # TODO: Integrate with resource operators and the new decomposition pipelines
+    """Return call graphs for given PennyLane Operator"""
+    return None
+
+
+@_get_op_call_graph.register
+def _(op: qtemps.subroutines.qpe.QuantumPhaseEstimation):
+    return {
+        qt_gates.Hadamard(): len(op.estimation_wires),
+        _map_to_bloq(op.hyperparameters["unitary"]).controlled(): (2 ** len(op.estimation_wires))
+        - 1,
+        _map_to_bloq((qtemps.QFT(wires=op.estimation_wires)), map_ops=False).adjoint(): 1,
+    }
+
+
+@singledispatch
+def _map_to_bloq(op, custom_mapping=None, map_ops=True, **kwargs):
+    """Map PennyLane operators to Qualtran Bloqs. Operators with direct equivalents are directly
+    mapped to their Qualtran equivalent even if ``map_ops`` is set to ``False``. Other operators are
+    given a smart default mapping. When given a ``custom_mapping``, the custom mapping is used."""
+
+    if custom_mapping is not None:
+        return custom_mapping[op]
+
+    return ToBloq(op, map_ops=map_ops, **kwargs)
+
+
+# pylint: disable=import-outside-toplevel
+@_map_to_bloq.register
+def _(
+    op: qtemps.subroutines.qpe.QuantumPhaseEstimation,
+    custom_mapping=None,
+    map_ops=True,
+    **kwargs,
+):
+    from qualtran.bloqs.phase_estimation import RectangularWindowState
+    from qualtran.bloqs.phase_estimation.text_book_qpe import TextbookQPE
```

Comment:
You could probably wrap these in an if statement for better error handling. Alternatively, you could just move the import above? 

```suggestion
    if not qualtran:
        raise ValueError("Can't map QPE to bloq without the qualtran dependency. Please install qualtran.")
    from qualtran.bloqs.phase_estimation import RectangularWindowState
    from qualtran.bloqs.phase_estimation.text_book_qpe import TextbookQPE
```

Reply:
Actually good point, but I think the error should be inside of qml.to_bloq() instead of here, right?